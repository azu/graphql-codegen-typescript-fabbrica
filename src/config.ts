import { ConvertFn, RawTypesConfig, convertFactory } from '@graphql-codegen/visitor-plugin-common';

export type RawConfig = {
  typesFile: string;
  skipTypename?: RawTypesConfig['skipTypename'];
  skipIsAbstractType?: boolean | undefined;
  namingConvention?: RawTypesConfig['namingConvention'];
  typesPrefix?: RawTypesConfig['typesPrefix'];
  typesSuffix?: RawTypesConfig['typesSuffix'];
  // TODO: support addIsAbstractType
};

export type Config = {
  typesFile: string;
  skipTypename: Exclude<RawTypesConfig['skipTypename'], undefined>;
  skipIsAbstractType: boolean;
  typesPrefix: Exclude<RawTypesConfig['typesPrefix'], undefined>;
  typesSuffix: Exclude<RawTypesConfig['typesSuffix'], undefined>;
  convert: ConvertFn;
  // TODO: support addIsAbstractType
};

export function validateConfig(rawConfig: unknown): asserts rawConfig is RawConfig {
  if (typeof rawConfig !== 'object' || rawConfig === null) {
    throw new Error('`options` must be an object');
  }
  if (!('typesFile' in rawConfig)) {
    throw new Error('`option.typesFile` is required');
  }
  if (typeof rawConfig['typesFile'] !== 'string') {
    throw new Error('`options.typesFile` must be a string');
  }
  if ('skipTypename' in rawConfig && typeof rawConfig['skipTypename'] !== 'boolean') {
    throw new Error('`options.skipTypename` must be a boolean');
  }
  if ('skipIsAbstractType' in rawConfig && typeof rawConfig['skipIsAbstractType'] !== 'boolean') {
    throw new Error('`options.skipIsAbstractType` must be a boolean');
  }
  if ('typesPrefix' in rawConfig && typeof rawConfig['typesPrefix'] !== 'string') {
    throw new Error('`options.typesPrefix` must be a string');
  }
  if ('typesSuffix' in rawConfig && typeof rawConfig['typesSuffix'] !== 'string') {
    throw new Error('`options.typesSuffix` must be a string');
  }
}

export function normalizeConfig(rawConfig: RawConfig): Config {
  return {
    typesFile: rawConfig.typesFile,
    skipTypename: rawConfig.skipTypename ?? false,
    skipIsAbstractType: rawConfig.skipIsAbstractType ?? true,
    typesPrefix: rawConfig.typesPrefix ?? '',
    typesSuffix: rawConfig.typesSuffix ?? '',
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    convert: convertFactory(rawConfig as any),
  };
}
