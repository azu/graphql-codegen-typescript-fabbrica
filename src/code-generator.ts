import { Config } from './config.js';
import {
    ExampleDirectiveFake,
    ExampleDirectivePattern,
    ExampleDirectiveValues,
    ObjectTypeInfo,
    TypeInfo
} from './schema-scanner.js';
import "@faker-js/faker"
function generatePreludeCode(config: Config, typeInfos: TypeInfo[]): string {
    const joinedTypeNames = typeInfos
        .filter(({ type }) => type === 'object')
        .map(({ name }) => `  ${name}`)
        .join(',\n');
    const code = `
import {
  type DefineTypeFactoryInterface${config.nonOptionalDefaultFields ? 'Required' : ''},
  defineTypeFactory,
} from '@mizdra/graphql-codegen-typescript-fabbrica/helper';
import type {
  Maybe,
${joinedTypeNames},
} from '${config.typesFile}';

export * from '@mizdra/graphql-codegen-typescript-fabbrica/helper';
  `.trim();
    return `${code}\n`;
}

export function generateOptionalTypeDefinitionCode(typeInfo: TypeInfo): string {
    if (typeInfo.type === 'object') {
        const { name, fields } = typeInfo;
        const comment = typeInfo.comment ?? '';
        const joinedPropDefinitions = fields
            .map((field) => {
                const comment = field.comment ? `  ${field.comment}` : '';
                return `${comment}  ${field.name}?: ${field.typeString};`;
            })
            .join('\n');
        return `
${comment}export type Optional${name} = {
${joinedPropDefinitions}
};
`.trimStart();
    } else {
        const { name, possibleTypes } = typeInfo;
        const comment = typeInfo.comment ?? '';
        const joinedPossibleTypes = possibleTypes.map((type) => `Optional${type}`).join(' | ');
        return `
${comment}export type Optional${name} = ${joinedPossibleTypes};
`.trimStart();
    }
}

function generateTypeFactoryCode({ nonOptionalDefaultFields }: Config, typeInfo: ObjectTypeInfo): string {
    const { name } = typeInfo;
    return `
/**
 * Define factory for {@link ${name}} model.
 *
 * @param options
 * @returns factory {@link ${name}FactoryInterface}
 */
export const define${name}Factory: DefineTypeFactoryInterface${nonOptionalDefaultFields ? 'Required' : ''}<
  Optional${name},
  {}
> = defineTypeFactory;
`.trimStart();
}

// export type ExampleDirectiveFake = {
//   fake: {
//     type: string
//     lang: string
//   }
// };
// export type ExampleDirectivePattern = {
//   pattern: string
// };
// export type ExampleDirectiveValues = {
//   values: (ValuePrimitive | ValueArray | ValueObject)[]
// };
const handleFake = (example: ExampleDirectiveFake) => {
    return `faker.${example.fake.type}()`;
};
const handlePattern = (example: ExampleDirectivePattern) => {
    return `faker.random.alphaNumeric(${example.pattern.length})`;
}
const handleValues = (example: ExampleDirectiveValues) => {
    return `faker.random.arrayElement(${example.values.map((value) => value).join(', ')})`;
}
const handleExample = (example: ExampleDirectiveFake | ExampleDirectivePattern | ExampleDirectiveValues) => {
    if ('fake' in example) {
        return handleFake(example);
    } else if ('pattern' in example) {
        return handlePattern(example);
    } else if ('values' in example) {
        return handleValues(example);
    }
    throw new Error('Unknown example directive');
}

export function generateDefaultFactory(_: Config, typeInfo: ObjectTypeInfo): string {
    const { name } = typeInfo;
    return `
/**
 * Factory for {@link ${name}} model.
 */
export const ${name}Factory = define${name}Factory({
  defaultFields: {
    ${typeInfo.fields
        // name: dynamic(() => faker.name.findName()),
        .flatMap(({ name,example }) => {
            if(!example) {
                return [];
            }
            return [`${name}: dynamic(() => ${handleExample(example)})`];
        })
        .join(',\n')}
  },
});`.trimStart();
}

export function generateCode(config: Config, typeInfos: TypeInfo[]): string {
    let code = '';
    code += generatePreludeCode(config, typeInfos);
    code += '\n';
    for (const typeInfo of typeInfos) {
        code += generateOptionalTypeDefinitionCode(typeInfo);
        code += '\n';
        if (typeInfo.type === 'object') {
            code += generateTypeFactoryCode(config, typeInfo);
            code += '\n';
            code += generateDefaultFactory(config, typeInfo);
            code += '\n';
        }
    }
    return code;
}
