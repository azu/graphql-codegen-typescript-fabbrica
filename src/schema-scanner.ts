import { getCachedDocumentNodeFromSchema } from '@graphql-codegen/plugin-helpers';
import { GraphQLObjectType, GraphQLSchema, visit } from 'graphql';
import { Config } from './config.js';
import { createTypeInfoVisitor } from './visitor.js';

// TODO: Support comment
type FieldInfo = { name: string; typeString: string };
export type TypeInfo = { name: string; fields: FieldInfo[] };

export function getTypeInfos(config: Config, schema: GraphQLSchema): TypeInfo[] {
  const userDefinedTypeNames = Object.values(schema.getTypeMap())
    // Ignore introspectionTypes
    // ref: https://github.com/graphql/graphql-js/blob/b12dcffe83098922dcc6c0ec94eb6fc032bd9772/src/type/introspection.ts#L552-L559
    .filter((type) => type instanceof GraphQLObjectType && !type.name.startsWith('__'))
    .map((type) => type.name);

  const visitor = createTypeInfoVisitor(config, userDefinedTypeNames);
  const ast = getCachedDocumentNodeFromSchema(schema);

  visit(ast, visitor);

  return visitor.getTypeInfos();
}
