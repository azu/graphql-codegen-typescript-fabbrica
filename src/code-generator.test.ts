import { describe, expect, it } from 'vitest';
import { generateCode } from './code-generator.js';
import { Config } from './config.js';
import { TypeInfo } from './schema-scanner.js';

describe('generateCode', () => {
  it('generates code', () => {
    const config: Config = {
      typesFile: './types',
    };
    const typeInfos: TypeInfo[] = [
      { name: 'Book', fieldNames: ['id', 'title', 'author'] },
      { name: 'Author', fieldNames: ['id', 'name', 'books'] },
    ];
    const actual = generateCode(config, typeInfos);
    expect(actual).toMatchSnapshot();
  });
});
