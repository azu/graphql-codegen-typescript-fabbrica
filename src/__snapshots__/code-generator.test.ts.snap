// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateCode > generates code 1`] = `
"import {
  type Traits,
  type TypeFactoryDefineOptions,
  type TypeFactoryInterface,
  type FieldsResolver,
  defineTypeFactoryInternal,
} from '@mizdra/graphql-codegen-typescript-fabbrica/helper';
import type {
  Maybe,
  Book,
  Author,
} from './types';

export * from '@mizdra/graphql-codegen-typescript-fabbrica/helper';

export type OptionalBook = {
  id?: string | undefined;
  title?: string | undefined;
  author?: OptionalAuthor | undefined;
};

export type BookFactoryDefineOptions<
  TransientFields extends Record<string, unknown>,
  _DefaultFieldsResolver extends FieldsResolver<OptionalBook & TransientFields>,
  _Traits extends Traits<OptionalBook, TransientFields>,
> = TypeFactoryDefineOptions<OptionalBook, TransientFields, _DefaultFieldsResolver, _Traits>;

export type BookFactoryInterface<
  TransientFields extends Record<string, unknown>,
  _DefaultFieldsResolver extends FieldsResolver<OptionalBook & TransientFields>,
  _Traits extends Traits<OptionalBook, TransientFields>,
> = TypeFactoryInterface<OptionalBook, TransientFields, _DefaultFieldsResolver, _Traits>;

/**
 * Define factory for {@link Book} model.
 *
 * @param options
 * @returns factory {@link BookFactoryInterface}
 */
export function defineBookFactory<
  _DefaultFieldsResolver extends FieldsResolver<OptionalBook>,
  _Traits extends Traits<OptionalBook, {}>,
>(
  options: BookFactoryDefineOptions<{}, _DefaultFieldsResolver, _Traits>,
): BookFactoryInterface<{}, _DefaultFieldsResolver, _Traits> {
  return defineTypeFactoryInternal([], options);
}

/**
 * Define factory for {@link Book} model with Transient Fields.
 *
 * @param defaultTransientFields
 * @returns defineTypeFactory {@link defineBookFactory}
 */
defineBookFactory.withTransientFields = <TransientFields extends Record<string, unknown>>(
  defaultTransientFields: TransientFields,
) => {
  return <
    _DefaultFieldsResolver extends FieldsResolver<OptionalBook & TransientFields>,
    _Traits extends Traits<OptionalBook, TransientFields>,
  >(
    options: BookFactoryDefineOptions<{}, _DefaultFieldsResolver, _Traits>,
  ): BookFactoryInterface<TransientFields, _DefaultFieldsResolver, _Traits> => {
    return defineTypeFactoryInternal(Object.keys(defaultTransientFields), {
      ...options,
      defaultFields: { ...(defaultTransientFields as FieldsResolver<TransientFields>), ...options.defaultFields },
    });
  };
};

export type OptionalAuthor = {
  id?: string | undefined;
  name?: string | undefined;
  books?: Book[] | undefined;
};

export type AuthorFactoryDefineOptions<
  TransientFields extends Record<string, unknown>,
  _DefaultFieldsResolver extends FieldsResolver<OptionalAuthor & TransientFields>,
  _Traits extends Traits<OptionalAuthor, TransientFields>,
> = TypeFactoryDefineOptions<OptionalAuthor, TransientFields, _DefaultFieldsResolver, _Traits>;

export type AuthorFactoryInterface<
  TransientFields extends Record<string, unknown>,
  _DefaultFieldsResolver extends FieldsResolver<OptionalAuthor & TransientFields>,
  _Traits extends Traits<OptionalAuthor, TransientFields>,
> = TypeFactoryInterface<OptionalAuthor, TransientFields, _DefaultFieldsResolver, _Traits>;

/**
 * Define factory for {@link Author} model.
 *
 * @param options
 * @returns factory {@link AuthorFactoryInterface}
 */
export function defineAuthorFactory<
  _DefaultFieldsResolver extends FieldsResolver<OptionalAuthor>,
  _Traits extends Traits<OptionalAuthor, {}>,
>(
  options: AuthorFactoryDefineOptions<{}, _DefaultFieldsResolver, _Traits>,
): AuthorFactoryInterface<{}, _DefaultFieldsResolver, _Traits> {
  return defineTypeFactoryInternal([], options);
}

/**
 * Define factory for {@link Author} model with Transient Fields.
 *
 * @param defaultTransientFields
 * @returns defineTypeFactory {@link defineAuthorFactory}
 */
defineAuthorFactory.withTransientFields = <TransientFields extends Record<string, unknown>>(
  defaultTransientFields: TransientFields,
) => {
  return <
    _DefaultFieldsResolver extends FieldsResolver<OptionalAuthor & TransientFields>,
    _Traits extends Traits<OptionalAuthor, TransientFields>,
  >(
    options: AuthorFactoryDefineOptions<{}, _DefaultFieldsResolver, _Traits>,
  ): AuthorFactoryInterface<TransientFields, _DefaultFieldsResolver, _Traits> => {
    return defineTypeFactoryInternal(Object.keys(defaultTransientFields), {
      ...options,
      defaultFields: { ...(defaultTransientFields as FieldsResolver<TransientFields>), ...options.defaultFields },
    });
  };
};

export type OptionalNode = OptionalBook | OptionalAuthor;

"
`;
